docs/appendices/data-types.txt

```diff
@@ -0,0 +1,176 @@
+__ https://crate.io/docs/crate/reference/en/latest/general/ddl/data-types.html#object
+__ https://docs.oracle.com/javase/8/docs/api/java/sql/JDBCType.html#JAVA_OBJECT
+__ https://crate.io/docs/crate/reference/en/latest/general/ddl/data-types.html#array
+__ https://docs.oracle.com/javase/8/docs/api/java/sql/JDBCType.html#ARRAY
+
+Array Types
+-----------
+
+Array types map to `java.sql.Array`_ instances.
+
+Use the ``.getArray()`` to get the underlying array (it is usually safe to
+cast it to ``Object[]``) if you prefer. Otherwise, to have JDBC handle type
+mapping, you can use ``.getResultSet()`` and use the related
+``ResultSetMetaData`` to get the array values converted to Java POJOs.
+
+The ``ResultSet`` will have one column with the inner array type and the name
+of the array field (``Array.getBaseType()``) and as many rows as there are
+elements in the array.
+
+Here's one example:
+
+FIXME: can an engineer please add explanatory comments to this block of code.
+please explain what is being done here. please also format the code in the most
+idiomatic yet reasonable way (I have given this a go myself) - please also add
+empty lines where it makes most sense
+
+.. code-block:: java
+
+    Statement statement = connection.createStatement();
+    ResultSet resultSet = statement.executeQuery(
+        "SELECT array_field FROM my_table"
+    );
+    resultSet.first();
+    Array arrayFieldArray = resultSet.getArray("array_field");
+    Object[] arrayFieldValue = arrayFieldValue.getArray();
+    ResultSet arrayFieldResultSet = arrayFieldArray.getResultSet();
+    arrayFieldResultSet.first();
+    String firstValue = arrayFieldResultSet.getString("array_field");
+
+When inserting arrays using a prepared statement, you must convert your array
+to a `java.sql.Array`_ by the use of ``createArrayOf()``. This function takes
+as its first argument, a CrateDB type as described above and as its second the
+array you want to convert.
+
+You can then use ``setArray()`` to set this converted array.
+
+FIXME: can an engineer please add explanatory comments to this block of code.
+please explain what is being done here. please also format the code in the most
+idiomatic yet reasonable way (I have given this a go myself)
+
+For example:
+
+.. code-block:: java
+
+    PreparedStatement preparedStatement = connection.prepareStatement(
+        "INSERT into my_table (string_array) VALUES (?)"
+    );
+    preparedStatement.setArray(1, connection.createArrayOf(
+        "string", new String[]{"a", "b"}
+    ));
+    preparedStatement.execute();
+
+Object Types
+------------
+
+Object columns map to a `java.util.Map<String, Object>`_ object.
+
+You can fetch them using ``ResultSet.getObject()`` and cast the result to
+``Map<String, Object>``. This ``Map`` will contain all nested columns defined in
+the object.
+
+.. CAUTION::
+
+   Objects can be ``null``.
+
+FIXIME: is the above right? which object? the resulting java object?
```

Comment:
please review this FIXME

Reply:
The sql object type can hold null values. In which case ``ResultSet.getObject()`` will return ``null``.